Bigquery insert

quartzmo opened the pull request:
This is an implementation of #1635 that does not precisely follow the suggested steps of first attempting to insert data, and then if the table is not found, creating the table and then reattempting the insert with backoff of >= 1 minute. Instead, if `autocreate` is true, it first performs a `Tables.get` to quickly check for the table, and if the table is not found, it creates the table and then does the insert. Our experience is that if the initial insert data call encounters a not found error, the service will continue to respond with the error for over one minute. See #1635 for discussion.

/cc @jskeet

blowmage commented on google-cloud-bigquery/acceptance/bigquery_helper.rb:
```diff
@@ -65,16 +65,16 @@ def setup
-      end
-      reporter.record result
-    end
+    # def self.run_one_method klass, method_name, reporter
```
Was this commented out for testing?

quartzmo replied:
Yes, thanks for catching this, I'll restore it.

tswast commented:
The backend team would prefer that `table.get` get called much less frequently than `tabledata.insertAll`. My understanding is that they propose the following logic:

1. Try to do `tabledata.insertAll` (get 404)
2. Wait a random delay (they didn't define the time, maybe between 1 and 60 seconds?).
3. Attempt to create the table.
4. Wait 1 minute.
5. Retry `tabledata.insertAll`

quartzmo commented:
@tswast And if 5. fails, repeat 4. and 5. indefinitely until 5. succeeds?

quartzmo commented:
Or:  if 5. fails, repeat 2., 3., 4. and 5. indefinitely until 5. succeeds?

Sorry, it's just tough to guess at the aggregate, real-world impact of these calls, so the more precisely you can spec it, the better. Anyway, I'll update this PR soon with a pass at it.

tswast commented:
My understanding is that we should repeat 2, 3, 4, and 5 on 404 errors, continuing indefinitely. Since this could potentially repeat forever, I'd love to see an optional timeout parameter.

quartzmo commented:
@tswast Thank you, SGTM. I'll update the PR tomorrow.

quartzmo commented:
I have updated the PR per the guidance above. The [acceptance test](https://github.com/quartzmo/google-cloud-ruby/blob/f2fc5d40a64b9939b4c8f46c71928f0b506efb4a/google-cloud-bigquery/acceptance/bigquery/dataset_test.rb#L274-L304) now takes several minutes to complete. (@blowmage, how do you feel about this additional time?) To show the sequence followed by the test, I added temporary debug logging statements that are now commented out, but need to be removed completely before accepting this PR. The output of my test runs is consistently as follows:

```sh
# Running:

NotFoundError rescue
sleeping before create_table
sleeping before insert
NotFoundError rescue
sleeping before create_table
AlreadyExistsError rescue
sleeping before insert
.

Finished in 205.142495s, 0.0049 runs/s, 0.1999 assertions/s.
```

@tswast I did not yet add the optional timeout parameter. Given the recursive behavior in the method, we believe it would require a separate thread, which will add some complexity and development time. Should we do it?

tswast commented:
Hmmm... An extra thread does sound more complex. Probably not worth it.

quartzmo commented:
@blowmage How do you feel about this  [acceptance test](https://github.com/quartzmo/google-cloud-ruby/blob/f2fc5d40a64b9939b4c8f46c71928f0b506efb4a/google-cloud-bigquery/acceptance/bigquery/dataset_test.rb#L274-L304) adding *two to three minutes* to the build?
